src/mono/wasm/runtime/hybrid-globalization/grapheme-segmenter.ts

```diff
@@ -0,0 +1,138 @@
+    return new RegExp(`${after ? "^" : ""}${replace_variables(variables, rule)}${after ? "" : "$"}`);
+}
+
+export class GraphemeSegmenter {
+    private readonly rules;
+    private readonly ruleSortedKeys;
+
+    public constructor() {  
+        // Process segmentation rules
+        this.rules = GraphemeSegmenter.prepare_segmanation_rules(SegmentationRules);
+        this.ruleSortedKeys = Object.keys(this.rules).sort((a, b) => Number(a) - Number(b));
+    }
+
+    /**
+     * Returns the next grapheme in the given string starting from the specified index.
+     * @param str - The input string.
+     * @param startIndex - The starting index.
+     * @returns The next grapheme.
+     */
+    public next_grapheme(str: string, startIndex: number): string {
+        const breakIdx = this.next_grapheme_break(str, startIndex);
+        return str.substring(startIndex, breakIdx);
+    }
+
+    /**
+     * Finds the index of the next grapheme break in a given string starting from a specified index.
+     * 
+     * @param str - The input string.
+     * @param startIndex - The index to start searching from.
+     * @returns The index of the next grapheme break.
+     */
+    public next_grapheme_break(str: string, startIndex: number): number {
+        if (startIndex < 0)
+            return 0;
+    
+        if (startIndex >= str.length - 1)
+            return str.length;
+    
+        let prev = String.fromCodePoint(str.codePointAt(startIndex)!);
+        for (let i = startIndex + 1; i < str.length; i++) {
+            // Don't break surrogate pairs
+            if (is_surrogate_pair(str, i)) {
+                continue;
+            }
+
+            const curr = String.fromCodePoint(str.codePointAt(i)!);
+            if (this.is_grapheme_break(prev, curr))
+                return i;
+    
+            prev = curr;
+        }
+    
+        return str.length;
+    }
+
+    private is_grapheme_break(previous: string, current: string): boolean {
+        for (const key of this.ruleSortedKeys) {
+            const {before, after, breaks} = this.rules[key];
+            // match before and after rules
+            if (before && !before.test(previous)) {
+                continue;
+            }
+            if (after && !after.test(current)) {
+                continue;
+            }
+
+            return breaks;
+        }
+
+        // GB999: Any ÷ Any
+        return true;
+    }
+
+    private static prepare_segmanation_rules(segmentationRules: SegmentationTypeRaw): Record<string, SegmentationRule> {
+        const preparedRules: Record<string, SegmentationRule> = {};
+    
+        for (const key of Object.keys(segmentationRules.rules)) {
+            const ruleValue = segmentationRules.rules[key];
+            const preparedRule: SegmentationRule = { breaks: ruleValue.breaks, };
+    
+            if ("before" in ruleValue && ruleValue.before) {
```

Comment:
can ruleValue.before or ruleValue.after ever be ""? If so, should we still generate regexes for those, or is it intended that we would not generate a regex in that case? (`""` in JS is falsy)

Reply:
If we know the rules won't ever be "" I think the current approach is fine.